### PR TITLE
dev-federation-fetch

### PR DESCRIPTION
**Pull request by rodrigolc**

### CODE REVIEW

File: federation/src/endpoint.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    spiffeid_ID spiffe_id;
+    spiffebundle_Source *bundle_source;
+    bool owns_bundle;
+} spiffebundle_Endpoint;
+
+spiffebundle_Endpoint *spiffebundle_Endpoint_New();
+void spiffebundle_Endpoint_Free(spiffebundle_Endpoint *endpoint);
+
+err_t spiffebundle_Endpoint_Config_HTTPS_WEB(
+    spiffebundle_Endpoint *endpoint, string_t url,
```
Comment (otaviolcs3) [inline federation/src/endpoint.h:38]:
```suggestion
    spiffebundle_Endpoint *endpoint, const char *url,
```

File: federation/src/endpoint.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    spiffeid_TrustDomain trust_domain);
+
+err_t spiffebundle_Endpoint_Config_HTTPS_SPIFFE(
+    spiffebundle_Endpoint *endpoint, string_t url,
```
Comment (otaviolcs3) [inline federation/src/endpoint.h:42]:
```suggestion
    spiffebundle_Endpoint *endpoint, const char *url,
```

File: bundle/spiffebundle/src/source.c [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        if(s->type == SPIFFEBUNDLE_BUNDLE) {
+            return spiffebundle_Bundle_GetBundleForTrustDomain(
+                s->source.bundle, td, err);
+        } else if(s->type == SPIFFEBUNDLE_SET) {
+            return spiffebundle_Set_GetBundleForTrustDomain(s->source.set, td,
+                                                            err);
+        } else if(s->type == SPIFFEBUNDLE_ENDPOINT) {
+            return spiffebundle_Endpoint_GetBundleForTrustDomain(
+                s->source.endpoint, td, err);
+        }
```
Comment (otaviolcs3) [inline bundle/spiffebundle/src/source.c:16]: Add 'else' case

File: federation/src/endpoint.c [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
+    curl_easy_setopt(curl, CURLOPT_URL, endpoint->url);
+    curl_easy_setopt(curl, CURLOPT_PORT, 443);
+    curl_easy_setopt(curl, CURLOPT_PROTOCOLS, CURLPROTO_HTTPS);
+    
+    
+    switch(endpoint->profile) {
+
+        int resp_code;
+    case HTTPS_WEB:
```
Comment (otaviolcs3) [inline federation/src/endpoint.c:239]: open braces here too

**Review by otaviolcs3 — COMMENTED** [review]
See comments

**Review by otaviolcs3 — APPROVED** [review]

File: federation/tests/resources/https_spiffe_server.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	// create a custom server with `TLSConfig`
+
+	s := &http.Server{
+		Addr:    "example.org:443",
+		Handler: handler, // use `http.DefaultServeMux`
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{*cert},
+			// GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+			// 	return cert, nil
+			// },
```
Comment (arianalima) [inline federation/tests/resources/https_spiffe_server.go:86]: Shouldn't it be de-commented or removed?

File: federation/tests/resources/https_spiffe_server.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		},
+	}
+
+	// run server on port "443"
+	go s.ListenAndServeTLS("", "")
+	// log.Fatal(s)
```
Comment (arianalima) [inline federation/tests/resources/https_spiffe_server.go:92]: same here

File: .github/workflows/actions.yml [hunk 7]
```diff
@@ -6,14 +6,26 @@ jobs:
...
         run: lsb_release -a
 
+      - name: Install dependencies
+        run: |
+          go version
+          go get -u github.com/spiffe/go-spiffe/v2/bundle/spiffebundle
+          go get -u github.com/spiffe/go-spiffe/v2/federation
+          go get -u github.com/spiffe/go-spiffe/v2/logger
+          go get -u github.com/spiffe/go-spiffe/v2/spiffeid
+          go get -u github.com/spiffe/go-spiffe/v2/svid/x509svid
```
Comment (arianalima) [inline .github/workflows/actions.yml:27]: Suggestion: add this steps in a bash script

File: federation/tests/resources/https_spiffe_server.go (hunk 5 above)
Comment (rodrigolc) [reply federation/tests/resources/https_spiffe_server.go:86]: fair enough!

File: federation/src/endpoint_example.c [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    spiffebundle_Endpoint *endpoint = spiffebundle_Endpoint_New();
+
+    // Configure Endpoint
+    // since raw.githubusercontent.com has a bundle installed for it, 
+    // the HTTPS call will authenticate.
+    spiffebundle_Endpoint_Config_HTTPS_WEB(
+        endpoint,
+        "https://raw.githubusercontent.com/HewlettPackard/c-spiffe/master/"
+        "bundle/jwtbundle/tests/resources/jwk_keys.json",
+        spiffeid_TrustDomainFromString("example.com", &err));
```
Comment (otaviolcs3) [inline federation/src/endpoint_example.c:17]: assign this trust domain to variable, so you can free it after

File: federation/src/endpoint_example.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    // Fetch it.
+    err = spiffebundle_Endpoint_Fetch(endpoint);
+
+    // Get it.
+    spiffebundle_Bundle *bundle
+        = spiffebundle_Endpoint_GetBundleForTrustDomain(
+            endpoint, spiffeid_TrustDomainFromString("example.com", &err),
```
Comment (otaviolcs3) [inline federation/src/endpoint_example.c:24]: Same here

File: .github/workflows/actions.yml [hunk 10]
```diff
@@ -72,12 +87,13 @@ jobs:
...
         shell: bash
-        run: cd build && make test
+        run: cd build && make test ARGS="--output-on-failure"
 
       - name: Coverage Test
         shell: bash
         run: cd build && make gcov && make lcov && lcov --list Coverage/lcov.info > coverage
-
+        
+      # TODO: re-enable when integration tests are fixed
```
Comment (arianalima) [inline .github/workflows/actions.yml:96]:
```suggestion
```

**Review by arianalima — APPROVED** [review]

File: federation/src/endpoint.c [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+    endpoint->url = NULL;
+    endpoint->profile = NONE;
+    endpoint->curl_handle = NULL;
+    return endpoint;
+}
+
+void spiffebundle_Endpoint_Free(spiffebundle_Endpoint *endpoint)
+{
+    if(endpoint) {
+        if(endpoint->owns_bundle) {
```
Comment (otaviolcs3) [inline federation/src/endpoint.c:49]: Get rid of the NULL check poitners

File: federation/src/endpoint.c [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+        return ERROR3;
+    }
+    endpoint->profile = HTTPS_WEB;
+    endpoint->owns_bundle = false;
+    return NO_ERROR;
+}
+
+err_t spiffebundle_Endpoint_Config_HTTPS_SPIFFE(
+    spiffebundle_Endpoint *endpoint, const char *url,
+    spiffeid_TrustDomain trust_domain, string_t spiffe_id,
```
Comment (otaviolcs3) [inline federation/src/endpoint.c:102]:
```suggestion
    spiffeid_TrustDomain trust_domain, const char *spiffe_id,
```

**Review by otaviolcs3 — CHANGES_REQUESTED** [review]
Some minor issues

**Review by otaviolcs3 — APPROVED** [review]

**Pull request merged**